packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,259 @@
+        address _target,
+        bytes memory _message
+    )
+        external
+        payable
+        nonReentrant
+    {
+        if (msg.sender != Predeploys.CROSS_L2_INBOX) revert RelayMessageCallerNotCrossL2Inbox();
+        if (CrossL2Inbox(Predeploys.CROSS_L2_INBOX).origin() != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
+            revert CrossL2InboxOriginNotL2ToL2CrossDomainMessenger();
+        }
+        if (_destination != block.chainid) revert MessageDestinationNotRelayChain();
+        if (_target == Predeploys.CROSS_L2_INBOX) revert MessageTargetCrossL2Inbox();
+        if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
+            revert MessageTargetL2ToL2CrossDomainMessenger();
+        }
+
+        bytes32 messageHash = keccak256(abi.encode(_destination, _source, _nonce, _sender, _target, _message));
+        if (successfulMessages[messageHash]) {
+            revert MessageAlreadyRelayed();
+        }
+
+        _storeMessageMetadata(_source, _sender);
+
+        successfulMessages[messageHash] = true;
+
+        bool success = _callWithAllGas(_target, _message);
+
+        if (success) {
+            emit RelayedMessage(messageHash);
+        } else {
+            successfulMessages[messageHash] = false;
+            emit FailedRelayedMessage(messageHash);
+        }
+
+        _storeMessageMetadata(0, address(0));
+    }
+
+    /// @notice Retrieves the next message nonce. Message version will be added to the upper two bytes of the message
+    ///         nonce. Message version allows us to treat messages as having different structures.
+    /// @return Nonce of the next message to be sent, with added message version.
+    function messageNonce() public view returns (uint256) {
+        return Encoding.encodeVersionedNonce(msgNonce, messageVersion);
+    }
+
+    /// @notice Retrieves whether the contract is currently entered or not.
+    /// @return True if the contract is entered, and false otherwise.
+    function entered() public view returns (bool) {
+        uint256 value;
+        assembly {
+            value := tload(ENTERED_SLOT)
+        }
+        return value != 0;
+    }
+
+    /// @notice Retrieves the sender of the current cross domain message.
+    /// @return _sender Address of the sender of the current cross domain message.
+    function _crossDomainMessageSender() internal view returns (address _sender) {
+        assembly {
+            _sender := tload(CROSS_DOMAIN_MESSAGE_SENDER_SLOT)
+        }
+    }
+
+    /// @notice Retrieves the source of the current cross domain message.
+    /// @return _source Chain ID of the source of the current cross domain message.
+    function _crossDomainMessageSource() internal view returns (uint256 _source) {
+        assembly {
+            _source := tload(CROSS_DOMAIN_MESSAGE_SOURCE_SLOT)
+        }
+    }
+
+    /// @notice Sets the entered value in transient storage to true.
+    function _setEnteredValueTrue() internal {
```

Comment:
I recommend just inlining these functions since they are only used once, ie `tstore` directly in the modifier, will let us delete a bunch of lines of code

Reply:
An alternative would be to have a single function that takes an arg instead of 2 functions and then call the func with diff args, but i think inlining still is better